### PR TITLE
Added a way to be forward compatible with version 2.0

### PR DESCRIPTION
**Pull request by Nyholm**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| Documentation   | 
| License         | MIT

To avoid updating all our plugins and the HTTPlug bundle, we can be forward compatible. 
https://3v4l.org/pn47M
https://3v4l.org/2kgtG

### CODE REVIEW

**Comment by dbu** [comment]
hm. i see what you want to do. but is it worth it? are the separate plugins that complicated? i'd rather tag a last 1.x version of those and then move to 2.x. there should be no other things depending on those plugins, so the cost of having a major version seems minimal to me.

**Comment by Nyholm** [comment]
HttplugBundle is also effected since it has a custom plugin and a custom client.

**Comment by GrahamCampbell** [comment]
I think this small BC layer is probably a good idea. Much less work for upgrading, and much less messy for trying to deal with 1.x and 2.x existing at the same time. Otherwise, we risk php-http ironically causing dependency conflicts and mess, with some people needing 1.x and others 2.x, purely because of the interface incompatibility of these 2 classes.

**Comment by Nyholm** [comment]
Traits are better. Sure!

**Comment by dbu** [comment]
looks good to me. pretty-ci seems to be a bit flaky these days :-(

**Comment by Nyholm** [comment]
No, it is our config file in master: https://github.com/php-http/client-common/blob/master/.php_cs

This is fixed in 2.x

File: src/Plugin/VersionBridgePlugin.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * A plugin that helps you migrate from php-http/client-common 1.x to 2.x. This
+ * will also help you to support PHP5 at the same time you support 2.x.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+trait VersionBridgePlugin
+{
+    abstract protected function doHandleRequest(RequestInterface $request, callable $next, callable $first);
+
+    public function handleRequest(RequestInterface $request, callable $next, callable $first)
```
Comment (xabbuh) [inline src/Plugin/VersionBridgePlugin.php]: Isn't this missing the return type hint?

File: src/VersionBridgeClient.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * A client that helps you migrate from php-http/httplug 1.x to 2.x. This
+ * will also help you to support PHP5 at the same time you support 2.x.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+trait VersionBridgeClient
+{
+    abstract protected function doSendRequest(RequestInterface $request);
+
+    public function sendRequest(RequestInterface $request)
```
Comment (xabbuh) [inline src/VersionBridgeClient.php]: same here

File: src/Plugin/VersionBridgePlugin.php (hunk 1 above)
Comment (Nyholm) [reply src/Plugin/VersionBridgePlugin.php]: This is for master (ie 1.9.0). 

Here we should be compatible with httplug 1. Check #133 for httplug 2.0.

Comment (xabbuh) [reply src/Plugin/VersionBridgePlugin.php]: Ah, I didn't see that PR yet. 👍

**Pull request merged**